libraries/AP_Math/tests/test_math.cpp

```diff
@@ -0,0 +1,309 @@
+{  
+  for(int i = 0; i < 1000; i++) {
+      if(i < 250) {
+          EXPECT_EQ(constrain_float(i, 250, 500), 250);
+          EXPECT_EQ(constrain_int16(i, 250, 500), 250);
+          EXPECT_EQ(constrain_int32(i, 250, 500), 250);
+      }
+      else if(i > 500) {
+          EXPECT_EQ(constrain_float(i, 250, 500), 500);
+          EXPECT_EQ(constrain_int16(i, 250, 500), 500);
+          EXPECT_EQ(constrain_int32(i, 250, 500), 500);
+      }
+      else {
+          EXPECT_EQ(constrain_float(i, 250, 500), i);        
+          EXPECT_EQ(constrain_int16(i, 250, 500), i);
+          EXPECT_EQ(constrain_int32(i, 250, 500), i);
+      }
+  }
+  
+  for(int i = 0; i <= 1000; i++) {
+      int c = i - 1000;
+      if(c < -250) {
+          EXPECT_EQ(constrain_float(c, -250, -50), -250);
+          EXPECT_EQ(constrain_int16(c, -250, -50), -250);
+          EXPECT_EQ(constrain_int32(c, -250, -50), -250);
+      }
+      else if(c > -50) {
+          EXPECT_EQ(constrain_float(c, -250, -50), -50);
+          EXPECT_EQ(constrain_int16(c, -250, -50), -50);
+          EXPECT_EQ(constrain_int32(c, -250, -50), -50);
+      }
+      else {
+          EXPECT_EQ(constrain_float(c, -250, -50), c);        
+          EXPECT_EQ(constrain_int16(c, -250, -50), c);
+          EXPECT_EQ(constrain_int32(c, -250, -50), c);
+      }
+  }
+  
+  for(int i = 0; i <= 2000; i++) {
+      int c = i - 1000;
+      if(c < -250) {
+          EXPECT_EQ(constrain_float(c, -250, 50), -250);
+          EXPECT_EQ(constrain_int16(c, -250, 50), -250);
+          EXPECT_EQ(constrain_int32(c, -250, 50), -250);
+      }
+      else if(c > 50) {
+          EXPECT_EQ(constrain_float(c, -250, 50), 50);
+          EXPECT_EQ(constrain_int16(c, -250, 50), 50);
+          EXPECT_EQ(constrain_int32(c, -250, 50), 50);
+      }
+      else {
+          EXPECT_EQ(constrain_float(c, -250, 50), c);        
+          EXPECT_EQ(constrain_int16(c, -250, 50), c);
+          EXPECT_EQ(constrain_int32(c, -250, 50), c);
+      }
+  }
+}
+
+TEST(Wrap180Test, angle180)
+{
+  // Full circle test
+  for(int i = 0; i < 360; i++) {
+      // smaller pole position
+      if(i < 180) {
+          EXPECT_EQ(wrap_180(i), i);
+          EXPECT_EQ(wrap_180(-i), -i);
+      }
+      // hit pole position -180/+180 degree
+      else if(i == 180) {
+          EXPECT_EQ(wrap_180(i), i);
+          EXPECT_EQ(wrap_180(-i), i);
+      }
+      // bigger pole position
+      else {
+          EXPECT_EQ(wrap_180(i), -(360-i));
+          EXPECT_EQ(wrap_180(-i), (360-i));
+      }
+  }
+  
+  EXPECT_EQ(wrap_180(45.f),        45.f);
+  EXPECT_EQ(wrap_180(90.f),        90.f);
+  EXPECT_EQ(wrap_180(180.f),       180.f);
+  EXPECT_EQ(wrap_180(180.1f),     -179.9f);
+  EXPECT_EQ(wrap_180(270.f),      -90.f);
+  EXPECT_EQ(wrap_180(360.f),       0.f);
+  EXPECT_EQ(wrap_180(720.f),       0.f);
+  EXPECT_EQ(wrap_180(3600.f),      0.f);
+  EXPECT_EQ(wrap_180(7200.f),      0.f);
+  EXPECT_EQ(wrap_180(-36000000.f), 0.f);
```

Comment:
Did you test it?
The value you are putting in the fmod function is losing precision over the argument due to addition
https://github.com/dgrat/ardupilot/blob/mathv7/libraries/AP_Math/AP_Math_Float.cpp#L90

Try this code ( with 4 byte float)

`#include <iostream>
#include <iomanip>
#include <cmath>
#include <limits>

template <class T>
bool is_zero(const T fVal) {
    static_assert(std::is_arithmetic<T>::value, "ERROR - is_zero(): template parameter not of type float or int\n");
    return fabsf(fVal) < std::numeric_limits<T>::epsilon() ? true : false; 
}

template <class T>
float wrap_360(const T &angle, float unit_mod = 1) {
    const float ang_360 = 360.f*unit_mod;
    float res = fmod(static_cast<float>(angle), ang_360);
    if (res < 0) {
        res += ang_360;
    }
    return res;
}

template <class T>
auto wrap_360_cd(const T &angle) -> decltype(wrap_360(angle, 100.f)) {
    return wrap_360(angle, 100.f);
}



template <class T>
float wrap_180(const T &angle, float unit_mod = 1) 
{   
    std::cout << "###### enter orig wrap_180 ##########################\n";
    const float ang_180 = 180.f*unit_mod;
    const float ang_360 = 360.f*unit_mod;
/*
      here I made the temporary in call to fmod an lvalue so we can see its value
*/
    float temp = static_cast<float>(angle) + ang_180;
    std::cout << "float temp           created in orig wrap_180 == " <<  std::setprecision(20) << temp << '\n';

    float res = fmod(temp, ang_360);

    // check using double
    double temp1 = static_cast<double>(angle) + ang_180;
    std::cout << "double version temp  created in orig wrap_180 == " <<  std::setprecision(20) << temp1 << '\n';

    if (res < 0 || is_zero(res)) {
        res += ang_180;
    }
    else {
        res -= ang_180;
    }
    std::cout <<  "###### leave orig wrap_180 ##########################\n\n";
    return res;
}


template <class T>
float wrap_180_V2(const T &angle, float unit_mod = 1) 
{ 
    auto res = wrap_360(angle, unit_mod);
    
    if ( res > T{18000}* unit_mod){
      res -= T{36000} * unit_mod;
    }
    return res;
}


template <class T>
float wrap_180_cd(const T &angle)
{
   return wrap_180(angle,100);
}

template <class T>
float wrap_180_cd_V2(const T &angle)
{
   return wrap_180_V2(angle,100);
}

int main()
{
   std::cout << "sizeof float == " << sizeof(float) << "\n\n";

   float x = 3600000000.f;

   float y_orig = wrap_180_cd(x);

   std::cout << "result of wrap_180_cd orig = "<<  std::setprecision(20)   << y_orig << "\n\n";


   float y_V2 = wrap_180_cd_V2(x);

   std::cout << "result of wrap_180_cd V2 = "<<  std::setprecision(20)   << y_V2 << "\n\n";
}
`